presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java

```diff
@@ -404,50 +404,66 @@ public synchronized void scheduleRequestIfNecessary()
         }
     }
 
-    public synchronized ListenableFuture<?> isBlocked()
+    public ListenableFuture<?> isBlocked()
     {
+        //  Fast path return without synchronizing
         if (isClosed() || isFailed() || pageBuffer.peek() != null) {
             return Futures.immediateFuture(true);
         }
-        SettableFuture<?> future = SettableFuture.create();
-        blockedCallers.add(future);
-        return future;
+        synchronized (this) {
+            //  Re-check after synchronizing
+            if (isClosed() || isFailed() || pageBuffer.peek() != null) {
+                return Futures.immediateFuture(true);
+            }
+            SettableFuture<?> future = SettableFuture.create();
+            blockedCallers.add(future);
+            return future;
+        }
     }
 
-    private synchronized boolean addPages(List<SerializedPage> pages)
+    private boolean addPages(List<SerializedPage> pages)
     {
+        //  Fast path return
         if (isClosed() || isFailed()) {
             return false;
         }
 
-        pageBuffer.addAll(pages);
+        //  Compute stats before acquiring the lock
+        long pagesRetainedSizeInBytes = 0;
+        long responseSize = 0;
+        for (SerializedPage page : pages) {
+            pagesRetainedSizeInBytes += page.getRetainedSizeInBytes();
+            responseSize += page.getSizeInBytes();
+        }
+
+        synchronized (this) {
+            //  Re-check after lock is acquired
+            if (isClosed() || isFailed()) {
+                return false;
+            }
+
+            pageBuffer.addAll(pages);
 
+            bufferRetainedSizeInBytes += pagesRetainedSizeInBytes;
+            maxBufferRetainedSizeInBytes = max(maxBufferRetainedSizeInBytes, bufferRetainedSizeInBytes);
+            successfulRequests++;
+            responseSizeExponentialMovingAverage.update(responseSize);
+            systemMemoryContext.setBytes(bufferRetainedSizeInBytes);
+        }
         if (!pages.isEmpty()) {
             // notify all blocked callers
             notifyBlockedCallers();
```

Comment:
Indeed it does, I think the best route here is going to be to clear the blocked callers list while holding the lock and notifying them after releasing it.

Reply:
In the common case pages will not be empty since one was just added so there is no reason to release the lock before acquiring the list of callers?

How common will it be that the list of blocked callers is non-empty? I imagine it depends a lot on the query and whether the producer or the consumer is faster.

Reply:
> In the common case pages will not be empty since one was just added so there is no reason to release the lock before acquiring the list of callers?

You're correct. Originally I restructured it this way to avoid holding the lock for the duration of the notifications being enqueued to the background executor (`ScheduledThreadPoolExecutorService` have a priority queue as their run queue even for non-scheduled runnables, so enqueues are more expensive than you might expect). I've updated the commit to just duplicate a little snippet of the `notifyBlockedCallers()` method with the synchronization in the right place.

> How common will it be that the list of blocked callers is non-empty? I imagine it depends a lot on the query and whether the producer or the consumer is faster.

It should depend on the relative pace of consumers + producers, but I think it's probably very common for the exchange network throughput is slower than whatever consumes from the other side of the exchange. As an aside, before the operator `ListenableFuture<?> isBlocked` caching change each driver would enqueue at least 3 separate listeners because of the way that the Driver loop discards the future on the first pass.